File: app/src/test/java/com/kareem/news_app/NewsViewModelTest.java
package com.kareem.news_app;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import kotlinx.coroutines.ExperimentalCoroutinesApi;

class NewsViewModelTest {
//    @ExperimentalCoroutinesApi
//
//    private lateinit var viewModel: ArticlesFragmentViewModel
//    private lateinit var useCase: GetArticlesUseCase
//    private lateinit var repository: ArticleRepository
//
//    @ExperimentalCoroutinesApi
//    @Before
//    fun setup() {
//        repository = object : ArticleRepository {
//            override suspend fun getArticles(pageNumber: Int): MutableList<ArticleDataModel> =
//                mutableListOf()
//        }
//        useCase = GetArticlesUseCase(repository)
//        viewModel = ArticlesFragmentViewModel(
//            useCase,
//            coroutinesTestRule.testCoroutineDispatcher
//        )
//    }
//
//    @Test
//    fun `test clear() then validate`() {
//        viewModel.clear()
//        assertTrue(viewModel.job.isCancelled)
//        assertTrue(!viewModel.isLoadingCheck)
//        assertTrue(viewModel.pageInfo.pageNumber == 1)
//        assertTrue(viewModel.pageInfo.hasNext)
//    }
//
//    @Test
//    fun `test onLoadMore() when page number is one expected result flow is Loading state`() {
//        viewModel.onLoadMore()
//        assertTrue(viewModel.articlesFlowObserver.value is ResultModel.Loading)
//    }
//
//    @Test
//    fun `test onLoadMore() when page number is greater than one expected result is not Loading state`() {
//        viewModel.pageInfo.pageNumber = 2
//        viewModel.onLoadMore()
//        assertTrue(viewModel.articlesFlowObserver.value !is ResultModel.Loading)
//    }
//
//    @Test
//    fun `test onLoadArticles() with success results expected result flow is Success state`() {
//        runBlocking {
//            viewModel.onLoadArticles(
//                ResultModel.Success(
//                    mutableListOf()
//                )
//            )
//        }
//
//        assertTrue(viewModel.articlesFlowObserver.value is ResultModel.Success)
//    }
//
//    @Test
//    fun `test onLoadArticles() with error results expected result flow is Error state`() {
//        runBlocking {
//            viewModel.onLoadArticles(
//                ResultModel.Error(
//                    null
//                )
//            )
//        }
//
//        assertTrue(
//            viewModel.articlesFlowObserver.value !is ResultModel.Success &&
//                    viewModel.articlesFlowObserver.value is ResultModel.Error
//        )
//    }
//
//    @Test
//    fun `test onLoadArticles() with success result and null data then page info has no next`() {
//        repository = object : ArticleRepository {
//            override suspend fun getArticles(pageNumber: Int): MutableList<ArticleDataModel>? = null
//        }
//
//        useCase = GetArticlesUseCase(repository)
//        runBlocking {
//            val response = useCase(1)
//            viewModel.onLoadArticles(response)
//            assertTrue(response is ResultModel.Success)
//            assertTrue(!viewModel.pageInfo.hasNext)
//        }
//    }
//
//    @Test
//    fun `test onLoadArticles() with success result and empty list then page info has no next`() {
//        useCase = GetArticlesUseCase(repository)
//        runBlocking {
//            val response = useCase(1)
//            viewModel.onLoadArticles(response)
//            assertTrue(response is ResultModel.Success)
//            assertTrue(!viewModel.pageInfo.hasNext)
//        }
//    }
//
//    @Test
//    fun `test onLoadArticles() with success result and not empty list then page info has next`() {
//        repository = object : ArticleRepository {
//            override suspend fun getArticles(pageNumber: Int): MutableList<ArticleDataModel> =
//                mutableListOf(
//                    ArticleDataModel(
//                        id = "1",
//                        "author1",
//                        content = "content1",
//                        description = "description1",
//                        publishedAt = "1234",
//                        title = "title1",
//                        "url1",
//                        "urlImage1"
//                    )
//                )
//        }
//        useCase = GetArticlesUseCase(repository)
//        runBlocking {
//            val response = useCase(1)
//            viewModel.onLoadArticles(response)
//            assertTrue(response is ResultModel.Success)
//            assertTrue(viewModel.pageInfo.hasNext)
//        }
    }

File: app/src/test/java/com/kareem/news_app/NewsViewModelTest.kt
package com.kareem.news_app

import com.kareem.domain.models.ArticleDataModel
import com.kareem.domain.repositories.NewsRepository
import com.kareem.domain.result.Resource
import com.kareem.domain.useCases.GetNewsUseCase
import com.kareem.news_app.view_models.NewsViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@ExperimentalCoroutinesApi
@RunWith(JUnit4::class)
internal class NewsViewModelTest {

    @get:Rule
    var coroutinesTestRule = CoroutinesTestRule()

    private lateinit var viewModel: NewsViewModel
    private lateinit var newsUseCase: GetNewsUseCase
    private lateinit var repository: NewsRepository

    @Before
    fun init() {
        repository = object : NewsRepository {
            override suspend fun getLatestNews(pageNumber: Int): MutableList<ArticleDataModel>? {
                return mutableListOf()
            }
        }
        newsUseCase = GetNewsUseCase(repository)
        viewModel = NewsViewModel(newsUseCase, coroutinesTestRule.testDispatcher)
    }

    @Test
    fun `test clearReferences() behavior then validate properties change`() {
        viewModel.clearReferences()
        assertTrue(!viewModel.isLoading)
        assertTrue(viewModel.pageInfo.pageNumber == 1)
        assertTrue(viewModel.pageInfo.hasNext)
        assertTrue(viewModel.job.isCancelled)
    }

    @Test
    fun `test onLoadMore() then validate state will change to loading`() {
        viewModel.onLoadMore()
        assertTrue(viewModel.newsStateFlow.value is Resource.Loading)
    }

    @Test
    fun `test onLoadNews() with success results then validate result has state flow is success`() {
        runBlocking {
            viewModel.onLoadNews(
                Resource.Success(
                    mutableListOf()
                )
            )
        }
        assertTrue(viewModel.newsStateFlow.value is Resource.Success)
    }

    @Test
    fun `test onLoadNews() with error results then validate result has state flow is success`() {
        runBlocking {
            viewModel.onLoadNews(
                Resource.Error(
                    "Error"
                )
            )
        }
        assertTrue(viewModel.newsStateFlow.value is Resource.Error)
    }

    @Test
    fun `test onLoadNews() with success result then validate page info has next`() {
        val mockRepositoryObject = object : NewsRepository {
            override suspend fun getLatestNews(pageNumber: Int): MutableList<ArticleDataModel> {
                return mutableListOf(
                    ArticleDataModel(
                        id = "1",
                        author = "jone Dao",
                        content = "content of Article",
                        description = "description of Article",
                        publishedAt = "2022-01-05T15:23:53Z",
                        title = "title of article",
                        sourceName = "source name of article",
                        url = "https://www.theverge.com/2022/1/5/22868388/meta-xros-vr-ar-os-development-canceled",
                        urlToImage = "https://cdn.vox-cdn.com/thumbor/iK-KTeHZYspkC6uhIxui7MfT2VY=/0x146:2040x1214/fit-in/1200x630/cdn.vox-cdn.com/uploads/chorus_asset/file/22977158/acastro_211101_1777_meta_0004.jpg"
                    )
                )
            }
        }
        val newsUseCase = GetNewsUseCase(mockRepositoryObject)
        runBlocking {
            val response = newsUseCase(1)
            viewModel.onLoadNews(response)
            assertTrue(response is Resource.Success)
            assertTrue(viewModel.pageInfo.hasNext)
        }
    }
    @Test
    fun `test onLoadNews() with success result then validate page info not has next`() {
        val mockRepositoryObject = object : NewsRepository {
            override suspend fun getLatestNews(pageNumber: Int): MutableList<ArticleDataModel>? = null
        }
        val newsUseCase = GetNewsUseCase(mockRepositoryObject)
        runBlocking {
            val response = newsUseCase(1)
            viewModel.onLoadNews(response)
            assertTrue(response is Resource.Success)
            assertTrue(viewModel.pageInfo.hasNext.not())
        }
    }
}